### PR TITLE
Use LLVM intrinsics for floating-point min/max

### LINKED ISSUE #18384: Use minnum/maxnum LLVM instrinsics for fmin/fmax
http://reviews.llvm.org/rL220341

Requires an LLVM upgrade, but not worth doing one just for this purpose.

### PR DESCRIPTION
**Pull request by varkor**
Resurrection of https://github.com/rust-lang/rust/pull/46926, now that the optimisation issues are fixed. I've confirmed locally that https://github.com/rust-lang/rust/pull/61384 solves the issues.

I'm not sure if we're allowed to move the `min`/`max` methods from libcore to libstd: I can't quite tell what the status is from https://github.com/rust-lang/rust/issues/50145. However, this is necessary to use the intrinsics.

Fixes https://github.com/rust-lang/rust/issues/18384.

r? @SimonSapin 
cc @rkruppe @nikic

### CODE REVIEW

**Comment by rkruppe** [comment]
This is a breaking change, the libcore methods are available on stable no_std crates.

**Comment by varkor** [comment]
It's bad that the breaking change doesn't break anything in our test suite. `no_std` should be tested more thoroughly.

What's the solution here, then? Do we need to resolve https://github.com/rust-lang/rust/issues/50145 somehow? It'd be good to work out a path to solving it.

File: src/libstd/f32.rs [hunk 1]
```diff
@@ -192,6 +192,38 @@ impl f32 {
...
+    /// let y = 2.0f32;
+    ///
+    /// assert_eq!(x.min(y), x);
+    /// ```
+    ///
+    /// If one of the arguments is NaN, then the other argument is returned.
+    #[stable(feature = "rust1", since = "1.0.0")]
+    #[inline]
+    pub fn min(self, other: f32) -> f32 {
+        unsafe { intrinsics::minnumf32(self, other) }
```
Comment (Centril) [inline src/libstd/f32.rs]: Do these need to be unsafe?

Comment (varkor) [reply src/libstd/f32.rs]: Intrinsics have to be unsafe.

Comment (Centril) [reply src/libstd/f32.rs]: Nope. See https://github.com/rust-lang/rust/blob/master/src/librustc_typeck/check/intrinsic.rs#L65 and https://github.com/rust-lang/rust/pull/57234/files.

Comment (varkor) [reply src/libstd/f32.rs]: Ah, I didn't realise this changed. These probably shouldn't need to be unsafe, then.

**Comment by SimonSapin** [comment]
*If* we are confident that these LLVM intrinsics never emit calls to functions that are expected to be provided by libc (but rather, when the target doesn’t have a dedicated instruction, falls back to emitting a sequence of "normal" instructions inline), then it is fine to use them in libcore.

If not, then yes this is a case of https://github.com/rust-lang/rust/issues/50145 and I don’t know how to start solving it.

Unfortunately I’ve had a hard time reading LLVM source code to be able to tell which intrinsics do what.

**Comment by nikic** [comment]
@SimonSapin These intrinsics can produce calls to fminf, fmin, fmaxf and fmax, at the very least for targets with hard float support.

It seems like we already have https://github.com/rust-lang-nursery/libm included in compiler-builtins, though it lacks these particular functions and I don't know what the state is of using that for non-wasm targets.

**Comment by varkor** [comment]
> It seems like we already have rust-lang-nursery/libm included in compiler-builtins, though it lacks these particular functions

Implementing these four functions is straightforward, at least.

> though it lacks these particular functions and I don't know what the state is of using that for non-wasm targets.

So, as I understand it, though LLVM [supports providing implementations](https://github.com/llvm-mirror/llvm/blob/29f026d977e01fc255284a0bb90afda1435b0e6a/include/llvm/IR/RuntimeLibcalls.def#L247-L256) of these intrinsics on platforms without native support, it's unclear whether rustc uses [rust-lang-nursery/libm](https://github.com/rust-lang-nursery/libm) on any platform but WASM.

However, [this milestone](https://github.com/rust-lang-nursery/libm/milestone/2) seems to indicate that math functions *are* used in core, though we should confirm this. If this is correct, then if we define fminf, fmin, fmaxf and fmax in the Rust libm, we should be good to use these intrinsics in core.

**Comment by varkor** [comment]
I've opened up a PR to add the min/max functions to libm: https://github.com/rust-lang-nursery/libm/pull/180. Once this is merged, the version of libm that compiler-builtins uses can be updated. After that, the version of compiler-builtins that core uses can be updated. Then, we should be good to use the intrinsics in core, as they'll be lowered properly by LLVM.
(This should also mean other intrinsics can be used in core, as long as they exist in compiler-builtins.)

**Comment by varkor** [comment]
@SimonSapin: this is now ready for re-review. This updates the version of compiler-builtins to one that includes the `fmin(f)`/`fmax(f)` libm function implementations, so LLVM should lower the intrinsics correctly on platforms that don't have native support. This means we should be able to use the LLVM intrinsics in libcore now.

**Comment by Centril** [comment]
cc @alexcrichton

**Comment by nikic** [comment]
To verify that the fallback to fmin() in no_std environment really works, maybe add a test with `+soft-float` target feature? Though might require disabling some other things like SSE to work on x86_64.

**Comment by SimonSapin** [comment]
Although I filed https://github.com/rust-lang/rust/issues/50145 it is only tracking the gap left by previous PRs and decisions. I don’t feel confident that I know the constraints and implications, so I’d prefer to defer this one.

r? @alexcrichton

**Comment by alexcrichton** [comment]
@bors: r+

Thanks @varkor!

We've already got a good suite of tests for the intrinsic implementations in libm in the libm repository itself, and we're arguably missing integration tests here in terms of making sure that the LLVM lowering is satisfied by compiler-builtins, but we typically handle those sorts of bugs over time as it's really difficult to write tests for them specificall here in this repository without too much CI burden.

**Pull request merged**